Support multiplication of doubles and TensorExpressions

macedo22 opened the pull request:
## Proposed changes

This PR creates a new `TensorExpression` type for storing `double`s, implements the multiplication of `double`s with `TensorExpression`s, and adds a test for these products.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 committed:
Implement scalar TensorExpression

macedo22 committed:
Support products of doubles and TensorExpressions

macedo22 committed:
Add test for products of doubles and TensorExpressions

macedo22 commented on src/DataStructures/Tensor/Expressions/Product.hpp:
```diff
@@ -233,3 +234,41 @@ SPECTRE_ALWAYS_INLINE auto operator*(
+  return t * TensorExpressions::ScalarAsExpression(s);
+}
+
+/// \ingroup TensorExpressionsGroup
```
Note that I copied and pasted the documentation from the `operator*` overload above because `\copydoc` seemed to have issues copying. My best guess for why it was not able to copy is because the template params for the `TensorExpression` operand are looked up with the `::` operator. This is what I used, which did not work:

```
/// \ingroup TensorExpressionsGroup
/// \copydoc operator*(const TensorExpression<T, double, typename
/// T::symmetry,typename T::index_list, ArgsList>&,const double)
```

I believe that if I move `T::symmetry` and `T::index_list` to default template parameters of the two new `operator*` overloads, then I think `\copydoc` would work:
```
/// documentation
/// ...
/// ...
template <typename T, typename ArgsList, typename Symmetry = T::symmetry, typename IndexList = T::index_list>
SPECTRE_ALWAYS_INLINE auto operator*(
    const TensorExpression<T, double, Symmetry, IndexList, ArgsList>& t,
    const double s) {
  return t * TensorExpressions::ScalarAsExpression(s);
}

/// \ingroup TensorExpressionsGroup
/// \copydoc operator*(const TensorExpression<T, double, Symmetry,
/// IndexList, ArgsList>&,const double)
template <typename T, typename ArgsList, typename Symmetry = T::symmetry, typename IndexList = T::index_list>
SPECTRE_ALWAYS_INLINE auto operator*(
    const double s,
   const TensorExpression<T, double, Symmetry, IndexList, ArgsList>& t) {
  return TensorExpressions::ScalarAsExpression(s) * t;
}
```

Let me know if this is preferred. The only benefit I see is not having to copy and paste the documentation. I didn't write it this way initially because I thought that a malformed call (where the symmetry of `t` is not the same as `T`) could override the default, but we have these default template parameter assignments elsewhere in the TE code, too.

nilsdeppe replied:
Have you tried:
```cpp
// @{
/// ...
auto operator*(T,s) ...

auto operator*(s,T)....
// @}
```
This `@{`-type syntax? I always get the order of the `@` and `}` wrong so you might need to look that up

nilsdeppe commented on src/DataStructures/Tensor/Expressions/ScalarAsExpression.hpp:
```diff
@@ -0,0 +1,54 @@
+
+namespace TensorExpressions {
+/// \ingroup TensorExpressionsGroup
+/// \brief Defines an expression representing a scalar
```
`scalar, e.g. a double`.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/ScalarAsExpression.hpp:
```diff
@@ -0,0 +1,54 @@
+  /// expression
+  /// \return the value of the scalar represented by this expression
+  template <typename Structure,
+            Requires<std::is_same_v<Structure, structure>> = nullptr>
```
since this isn't being used for overload selection, do you think having a `static_assert` enforcing this in the body would be better? Then instead of `Can't find function...` the error message is what you write in the static assert

macedo22 replied:
Sure, I'll do that instead!

nilsdeppe commented on src/DataStructures/Tensor/Expressions/ScalarAsExpression.hpp:
```diff
@@ -0,0 +1,54 @@
+
+ private:
+  // The scalar value being represented as an expression
+  double s_;
```
maybe rename `s_` to `scalar_value_` and then you don't need the comment since the code is (hopefully) clear at that point?

macedo22 replied:
I also updated the `operator*` overloads with scalars to use `scalar` instead of `s` for the argument name in a fixup commit so that the consistency keeps the connection between the `operator*` argument and `ScalarAsExpression` member clear.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/Product.hpp:
```diff
@@ -233,3 +234,41 @@ SPECTRE_ALWAYS_INLINE auto operator*(
+    const TensorExpression<T, double, typename T::symmetry,
+                           typename T::index_list, ArgsList>& t,
+    const double s) {
+  return t * TensorExpressions::ScalarAsExpression(s);
```
can you instead make this `ScalarAsExpression(s) * t`? I think we discussed having the scalars on the LHS at some point

macedo22 replied:
I considered doing this switch originally and then I didn't because - being completely honest - although I felt convinced it would matter when we briefly discussed it, in tracing through the syntax tree in my head, I didn't really see what the practical difference would be. Since I didn't see a perceived difference, I left the original order from the user intact because then the generated tree matches what they wrote and would be easier to debug.

If there would be a practical difference in permuting these operands, happy to change it! If so, could you provide a case where you feel like reordering would affect optimization/accuracy? I'm generally aware that multiplication order can affect the end result, but wasn't sure what the benefit of doing this switch would be. Let me know!

nilsdeppe reviewed (commented):
you can squash the fixups immediately :)

macedo22 commented:
@nilsdeppe For some reason I'm unable to reply directly to the doxygen suggestion, but that worked! I had seen those `// @{` and `// @}` in other parts of the code, but never knew what it did. Pretty cool!

So far, I think everything is addressed outside of the operand swapping suggestion. Pending your response to that and a potential fixup for it, I'll squash the fixups.

nilsdeppe reviewed (commented):
Okay, let's leave that as-is. It's easy to change if profiling data ultimately says we should change the order :)

Please go ahead squash the fixups you have :)

macedo22 commented:
@nilsdeppe I decided against swapping the ordering because I think even in the cases like

**1.**
```
Tensor<DataVector, indices....> * double * double
```
```
double * Tensor<DataVector, indices....> * double * double
```

I think requesting people rewrite manually like
**2.**
```
double * double * Tensor<DataVector, indices....>
```
```
double * double * double * Tensor<DataVector, indices....>
```

would still be wanted even if we did have the `operator*` permute the operands in the tree at this level.

To demonstrate the difference, I believe the TEs generated for **1.** are
```
(with the swap, swaps the Scalar and Tensor at the lowest level)
                            OuterProduct
                          /              \
                OuterProduct             ScalarAsExpression
                  /        \
ScalarAsExpression    TensorAsExpression
```
```
(I think should be the same whether or not we swap the operands in this return statement)
                                   OuterProduct
                                  /           \
                            OuterProduct      ScalarAsExpression
                          /              \
                OuterProduct             ScalarAsExpression
                  /        \
ScalarAsExpression    TensorAsExpression
```
whereas the TEs for **2.** I think both are:
```
            OuterProduct
            /          \
ScalarAsExpression TensorAsExpression
```


So I think requiring people write the scalars in the front manually is probably something we would want to enforce because it saves on time and space (less `get` traversing, smaller TEs), and who knows what other things the compiler might be able to optimize by having the `double`s in the expression (not wrapped in a `ScalarAsExpression`) multiplied before multiplying that result with a retrieved value from an expression.

macedo22 commented on src/DataStructures/Tensor/Expressions/Product.hpp:
```diff
@@ -56,14 +61,16 @@ template <typename T1, typename T2, template <typename...> class IndexList1,
           typename detail::OuterProductType<T1, T2>::index_list,
           typename detail::OuterProductType<T1, T2>::tensorindex_list> {
-  static_assert(std::is_same<typename T1::type, typename T2::type>::value,
+  static_assert(std::is_same<typename T1::type, typename T2::type>::value or
```
This should mean we support:
-`double * TE<double, indices....>`
-`double * TE<DataVector, indices....>`
Meaning, `double * (any derived TE, not just TensorAsExpression, whose type is double or DataVector)`

But is this also supposed to be supported?:
`TensorAsExpression<double> * TE<DataVector, indices...>`
Meaning, `(TensorAsExpression storing a pointer to a rank 0 Tensor storing a double) * (any derived TE, not just TensorAsExpression, whose type is double or DataVector)`. Example:
- `R() * T(ti_a, ti_b)`, where `R` is a rank 0 `Tensor` storing a double, and `T` is a rank 2 `Tensor` storing `DataVector`s

nilsdeppe replied:
No, I think only `double * TE` should be supported. I don't really see how `R() * TE<DataVector>` with R being a double is useful. I'd expect the double to be some constant, e.g. 0.5 or 0.25, or we are multiplying by an entire field.  The more general case just means relaxing a static assert, right? If that's the case I'd say implement the restricted case, we can generalize easily later if necessary

macedo22 replied:
Sounds good!

fmahebert commented on src/DataStructures/Tensor/Expressions/ScalarAsExpression.hpp:
```diff
@@ -0,0 +1,54 @@
+  static constexpr auto num_tensor_indices = 0;
+
+  /// \brief Create an expression from a scalar
+  ScalarAsExpression(double scalar) : scalar_(scalar) {}
```
`const double` I think

fmahebert reviewed (commented):
Are we concerned that the introduction of a new meaning of "scalar" could be confusing? We already have a `Scalar<DataType>` type alias for a rank-0 tensor. I'm wondering if using `DoubleAsExpression` here instead would help... I'm not sold on this suggestion, so that's not a change request... thoughts?

macedo22 commented:
@fmahebert After recently learning about `Scalar<DataType>`, I think you could be right. I actually initially had this new expression type called `DoubleAsExpression` at first and then changed it. I'm happy to change it back since you also thought of this! @nilsdeppe do you have opinions on this before I make this change? I think you two know better how problematic `ScalarAsExpression` would be considering the usage of `Scalar<DataType>`.

nilsdeppe commented:
What about `FundamentalAsExpression`? I'm thinking ahead to when someone wants complex numbers :) The rest of tenex I think supports complex numbers, this would be the first place where `double`/real numbers are explicitly assumed. I don't think we need to immediately add support and tests for complex numbers, but one name change is easier than multiple, and then just document that support for more fundamental types could be added. Does that sound okay?

macedo22 commented:
I'll defer to @fmahebert since I am not as aware of how broadly complex numbers or the term `Fundamental` is used in naming other things in SpECTRE or GR. I'm not personally aware of the umbrella term for numbers that include both real and complex. (Aside: technically aren't all real numbers also complex? `NumberAsExpression` seems maybe too vague and `ComplexNumberAsExpression` would be misleading, since I don't think most people would naturally see that name and think it also includes plain old real number `double`s). I'm happy with `DoubleAsExpression` or `FundamentalAsExpression`, depending on what you two think

nilsdeppe commented:
By `Fundamental` I meant a fundamental type. While complex numbers aren't _technically_ fundamental, they act like it in pretty much all cases

macedo22 commented:
Ah, I see. I can see how complex numbers would just be represented by two fundamental values in code, e.g. `{1.0, 5.0}` for 1 + 5i or something.

If we're wanting to think ahead with naming to include complex numbers, then I might slightly prefer `NumberAsExpression` over `FundamentalAsExpression` just because it seems like "Number" can include non-complex number values (makes me think `double`, `size_t`, `int`) and complex number values, while "Fundamental," to me, makes me think `size_t`, `int`, `char`, `bool`, `double`, etc (even though `char` and `bool` are also numbers) and I don't think complex values sit under "fundamental type" as naturally as they do under "number." So while `NumberAsExpression` seems a little vague, it also seems maybe a bit more fitting in capturing what we want represented as an expression (`double`s and complex values).

That's just my two cents, but I can hear myself getting annoyingly semantic :laughing:  I'm perfectly happy with whatever we call it as long as you and @fmahebert think that it would be most intuitive for SpeCTRE developers. I don't have a strong preference on this and think it's better for you two to make the call over me, and then I will fulfill my duty in renaming to whatever is decided :+1:

nilsdeppe commented:
`NumberAsExpression` is fine with me

fmahebert commented:
`NumberAsExpression` sounds good to me, I think it's pretty unambiguous :)

macedo22 commented:
@nilsdeppe Something I just thought of since you said the other TEs should work with complex numbers - what would this mean for the `type` aliases defined in each TE class? Currently, there are some checks in place in different TEs that check that the data types of operands are the same, e.g. [here](https://github.com/sxs-collaboration/spectre/blob/5234f36203a5c05837c146a449e1e52aa9716b9d/src/DataStructures/Tensor/Expressions/AddSubtract.hpp#L64) in `AddSub`. Sorry if this is obvious! Never worked with complex numbers in code before. I imagine supporting complex numbers might require tweaking checks like this.

nilsdeppe commented:
Good question! Getting complex numbers to work correctly would probably be a bit of work, since those checks should really be that `T1::type +/- T2::type` is a valid expression. I think this can be done with expression SFINAE (think `decltype`). Maybe something like
```cpp
template <typename A, typename B, typename = std::void_t<>>
struct CanAdd : std::false_type {};

template <typename A, typename B>
struct CanAdd<A, B, std::void_t<decltype(std::declval<A>() + std::declval<B>())>> : std::true_type{};
```

macedo22 commented:
Interesting! TIL `std::declval`. I imagine that the above expression SFINAE would also be on the pretense that we have a `operator+` overload defined that would accept `TE<double> + TE<std::complex<double>>` or something like that? Basically, wouldn't the `operator+` overload that accepts what we want need to be defined for that expression SFINAE to work? I don't have much experience with SFINAE, so just checking my understanding

nilsdeppe commented:
Yes, it would basically "recurse" through the types to check things are valid, though I think `::type` is always a `DataVector` (or `ComplexDataVector`) expression (or a `double/std::complex<double>`)

macedo22 commented:
Sounds good!

macedo22 commented:
@nilsdeppe @fmahebert I've squashed the old fixup requests, and in the new fixups I updated naming for each of the commits. Let me know if you need me to rename the message of the first commit if "scalar" is confusing/misleading

fmahebert commented:
LGTM 👍

nilsdeppe commented:
LGTM

macedo22 commented:
Squashed and rebased!

nilsdeppe commented on tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp:
```diff
@@ -33,6 +33,53 @@ void assign_unique_values_to_tensor(
   }
 }
 
+/// \ingroup TestingFrameworkGroup
```
This is in a cpp file, so don't need/use a Doxygen comment (two slashes, no `\ingroup`). I'm generally fine with formatting the comment in the Doxygen style since that is actually reasonably easy to read compared to when everyone comes up with their own comment style :P

This is true for the other functions in this file (clearly I missed that in other PRs), so please just clean that up in a separate PR

macedo22 replied:
Oh I didn't realize that doxygen documentation wasn't generated for cpp files (clearly, I haven't been thorough enough in checking all of my documentation, which is on me). So I'm clear, you're saying you're fine with even keeping the doxygen commands/labels like `\brief`, `\details`, `\tparam`, `\param`, `\return`?

nilsdeppe replied:
Yes, that's right
